src/cuda/backend.f90

```diff
@@ -415,21 +421,43 @@ subroutine tds_solve_dist(self, du, u, dirps, tdsops, blocks, threads)
 
    end subroutine tds_solve_dist
 
-   subroutine trans_x2y_cuda(self, u_y, v_y, w_y, u, v, w)
+   subroutine trans_x2y_cuda(self, u_y, u)
       implicit none
 
       class(cuda_backend_t) :: self
-      class(field_t), intent(inout) :: u_y, v_y, w_y
-      class(field_t), intent(in) :: u, v, w
+      class(field_t), intent(inout) :: u_y
+      class(field_t), intent(in) :: u
+
+      real(dp), device, pointer, dimension(:, :, :) :: u_d, u_y_d
+      type(dim3) :: blocks, threads
+
+      select type(u); type is (cuda_field_t); u_d => u%data_d; end select
```

Comment:
WRT comment that "select type stuff is annoying" - I agree. One possible way to make it slightly better would be
```
subroutine set_data_pointer(u, u_d)
  class(field_t), intent(in) :: u
  real(dp), device, pointer, dimension(:, :, :), intent(out) :: u_d
  select type(u)
  type is (cuda_field_t)
    u_d => u%data_d
  end select
end subroutine
```
and then the code in the main subroutine should be
```
call set_data_pointer(u, u_d)
call set_data_pointer(u_y, u_y_d)
```
which is perhaps a little better.

The other option would be to rewrite this as a wrapper subroutine accepting `class(field_t)` and then it calls the actual subroutine with `type(cuda_field_t)` arguments. The wrapper subroutine would need to do the select type dance (and should probably use `class default` for some error checking), but the actual implementation would be free if this complication which may or may not be an improvement.

Reply:
I think we need something like this for sure, but was planning to deal with it later on when fixing the allocator.  The issue is mentioned here #24. And I think we can pass the size we need into this `set_data_pointer` function and handle the bounds remapping there.

Reply:
I agree, better not to clutter the change.